### PR TITLE
test(rust): ensure deterministic proptests

### PR DESCRIPTION
**Pull request by thomaseizinger**
For quite a while now, we have been making extensive use of property-based testing to ensure `connlib` works as intended. The idea of proptests is that - given a certain seed - we deterministically sample test inputs and assert properties on a given function.

If the test fails, `proptest` prints the seed which can then be added to a regressions file to iterate on the test case and fix it. It is quite obvious that non-determinism in how the test input gets generated is no bueno and reduces the value we get out of these tests a fair bit.

The `HashMap` and `HashSet` data structures are known to be non-deterministic in their iteration order. This causes non-determinism during the input generation because we make use of a lot of maps and sets to gradually build up the test input. We fix all uses of `HashMap` and `HashSet` by replacing them with `BTreeMap` and `BTreeSet`.

To ensure this doesn't regress, we refactor `tunnel_test` to not make use of proptest's macros and instead, we initialise and run the test ourselves. This allows us to dump the sampled state and transitions into a file per test run. In CI, we then run a 2nd iteration of all regression tests and compare the sampled state and transitions with the previous run. They must match byte-for-byte.

Finally, to discourage use of non-deterministic iteration, we ban the use of the iteration functions on `HashMap` and `HashSet` across the codebase. This doesn't catch iteration in a `for`-loop but it is better than not linting against it at all.

### CODE REVIEW

File: .github/workflows/_rust.yml
Comment (ReactorScram) [inline .github/workflows/_rust.yml:120]: If this is not set, it won't save any files, right? Just wondering if I should be keeping track of test runs on my dev systems.

File: rust/clippy.toml
Comment (ReactorScram) [inline rust/clippy.toml]:
```suggestion
    { path = "std::collections::HashMap::iter", reason = "HashMap has non-deterministic iteration order, use BTreeMap instead" },
    { path = "std::collections::HashSet::iter", reason = "HashSet has non-deterministic iteration order, use BTreeSet instead" },
    { path = "std::collections::HashMap::into_iter", reason = "HashMap has non-deterministic iteration order, use BTreeMap instead" },
    { path = "std::collections::HashSet::into_iter", reason = "HashSet has non-deterministic iteration order, use BTreeSet instead" },
```

File: rust/connlib/clients/android/src/lib.rs
Comment (ReactorScram) [inline rust/connlib/clients/android/src/lib.rs:487]: Do we iterate over the list of disabled resources? 🤔

File: rust/connlib/snownet/src/node.rs
Comment (ReactorScram) [inline rust/connlib/snownet/src/node.rs]:
```suggestion
    host_candidates: Vec<Candidate>, // `Candidate` doesn't implement `PartialOrd` so we cannot use a `BTreeSet`. Linear search is okay because we expect this vec to be <100 elements
```

**Review by ReactorScram — COMMENTED** [review]
This bit me in Lua years ago and I didn't even think to look for it.

What if we forbid HashMap and HashSet entirely and then allow a wrapper that forwards only deterministic methods?

Cause I'm wondering how many of these `BTreeMap` substitutions really get iterated over for anything except debug printing.

File: .github/workflows/_rust.yml
Comment (thomaseizinger) [reply .github/workflows/_rust.yml:120]: It will always save files but they are gitignored. Could make it conditional on the env var yeah.

File: rust/connlib/clients/android/src/lib.rs
Comment (thomaseizinger) [reply rust/connlib/clients/android/src/lib.rs:487]: We don't but they appear in `Transition` and thus my non-determinism detection picked it up.

Comment (thomaseizinger) [reply rust/connlib/clients/android/src/lib.rs:487]: We do when we process them in connlib I think. It is deterministic without this too but the list appears in `Transition` and hence my non-determinism detection got set off.

**Comment by thomaseizinger** [comment]
> What if we forbid HashMap and HashSet entirely and then allow a wrapper that forwards only deterministic methods?

clippy can only ban methods, not types I believe. And using `HashMap` is fine unless you iterate over them somewhere in connlib or in the proptests.

**Comment by thomaseizinger** [comment]
This also fixes a bug in the proptests so if CI on `main` gets annoying because it is flaky, feel free to merge early.

File: .github/workflows/_rust.yml
Comment (ReactorScram) [reply .github/workflows/_rust.yml:120]: Okay. Sometimes I get odd failures locally and I'm not sure how to replicate / report them, so I'll try uploading the files if that happens after this PR merges

**Review by ReactorScram — APPROVED** [review]
Approving to unblock but I do see a `disallowed-types` for Clippy here: https://rust-lang.github.io/rust-clippy/master/index.html#/disallowed_types

Funny enough the example is the opposite, banning `BTreeMap`

**Comment by jamilbk** [comment]
Ok if I merge this? I seem to be hitting it on each PR now.

**Comment by thomaseizinger** [comment]
> Approving to unblock but I do see a `disallowed-types` for Clippy here: https://rust-lang.github.io/rust-clippy/master/index.html#/disallowed_types
> 
> Funny enough the example is the opposite, banning `BTreeMap`

Ah interesting! We can check how invasive that is. It would be useful to only only ban the types in certain directories.

See https://github.com/rust-lang/rust-clippy/issues/7353.

**Comment by thomaseizinger** [comment]
> What if we forbid HashMap and HashSet entirely and then allow a wrapper that forwards only deterministic methods?

Yeah that would work! The wrappers will be quite a bit of code and we need an extra crate like `stdext` or something.

I am curious to see how that would turn out. Want to have a go at it? :)

**Comment by thomaseizinger** [comment]
Uuh it found another bug.

**Comment by ReactorScram** [comment]
Yeah I could try it next week if I need a break from iOS favorites menu. I made an issue in the backlog here https://github.com/firezone/firezone/issues/6331

**Pull request merged**
